include/ppx/knob.h

```diff
@@ -0,0 +1,389 @@
+    : public Knob
+{
+public:
+    KnobDropdown(const std::string& flagName, size_t defaultIndex, std::vector<T>::iterator choicesBegin, std::vector<T>::iterator choicesEnd)
+        : Knob(flagName)
+    {
+        PPX_ASSERT_MSG(choicesBegin != choicesEnd, "too few allowed choices");
+        PPX_ASSERT_MSG(static_cast<int>(defaultIndex) > 0 && static_cast<int>(defaultIndex) < choicesEnd - choicesBegin, "defaultIndex is out of range");
+
+        mChoices = std::vector<T>(choicesBegin, choicesEnd);
+        SetDefault(defaultIndex);
+    }
+
+    void Draw() override
+    {
+        if (ImGui::BeginCombo(mDisplayName.c_str(), mChoices.at(mIndex).c_str())) {
+            for (int i = 0; i < mChoices.size(); ++i) {
+                bool isSelected = (i == mIndex);
+                if (ImGui::Selectable(mChoices.at(i).c_str(), isSelected)) {
+                    if (i != mIndex) { // A new choice is selected
+                        mIndex = i;
+                    }
+                }
+                if (isSelected) {
+                    ImGui::SetItemDefaultFocus();
+                }
+            }
+            ImGui::EndCombo();
+        }
+    }
+
+    void ResetToDefault() override { SetIndex(mDefaultIndex); }
+
+    // GetFlagHelpText format:
+    // --flag_name <"choice1"|"choice2"|...> : flag description
+    std::string GetFlagHelpText() const override
+    {
+        std::string choiceStr = "";
+        for (auto choice : mChoices) {
+            choiceStr += '\"' + choice + '\"' + "|";
+        }
+        if (!choiceStr.empty())
+            choiceStr.erase(choiceStr.size() - 1);
+
+        std::string flagHelpText = "--" + mFlagName + " <" + choiceStr + ">";
+        if (!mFlagDesc.empty()) {
+            flagHelpText += " : " + mFlagDesc;
+        }
+        return flagHelpText + "\n";
+    }
+
+    // Updates mIndex and mDefaultIndex, expected flag format:
+    // --flag_name <str>
+    void UpdateFromFlag(const CliOptions& opts) override
+    {
+        std::string newVal = opts.GetExtraOptionValueOrDefault(mFlagName, GetStr());
+        SetDefault(newVal);
+    }
+
+    size_t      GetIndex() const { return mIndex; }
+    std::string GetStr() const { return mChoices[mIndex]; }
+
+    // Used for when mIndex needs to be updated outside of UI
+    // Updates mIndex
+    void SetIndex(size_t newI)
+    {
+        if (!IsValidIndex(newI)) {
+            PPX_LOG_ERROR(mFlagName << " does not have this index in allowed choices: " << newI);
+            return;
+        }
+
+        mIndex = newI;
+    }
+
+    // Used for setting from flags
+    // Updates mIndex
+    void SetIndex(std::string newVal)
+    {
+        auto temp = std::find(mChoices.begin(), mChoices.end(), newVal);
+        if (temp == mChoices.end()) {
+            PPX_LOG_ERROR(mFlagName << " does not have this value in allowed range: " << newVal);
+            return;
+        }
+
+        mIndex = static_cast<size_t>(temp - mChoices.begin());
```

Comment:
`std::distance(mChoices.begin(), temp)`
Also nit: if you don't need to write, you can use `cbegin()`/`cend()`